File: windows/agora_render_performance_monitor.cc
#include "include/agora_rtc_engine/agora_render_performance_monitor.h"

#include <chrono>
#include <cmath>
#include <numeric>
#include <thread>

namespace agora {
namespace rtc {
namespace flutter {

/*
std::map<std::string, double> AgoraRenderPerformanceStats::toDictionary() const {
    std::map<std::string, double> dict;
    dict["uid"] = static_cast<double>(uid);
    dict["renderInputFps"] = renderInputFps;
    dict["renderOutputFps"] = renderOutputFps;
    dict["renderFrameIntervalMs"] = renderFrameIntervalMs;
    dict["renderDrawCostMs"] = renderDrawCostMs;
    dict["totalFramesReceived"] = static_cast<double>(totalFramesReceived);
    dict["totalFramesRendered"] = static_cast<double>(totalFramesRendered);
    return dict;
}
*/

AgoraRenderPerformanceMonitor::AgoraRenderPerformanceMonitor()
    : delegate_(nullptr),
      enabled_(true),
      totalFramesReceived_(0),
      totalFramesRendered_(0) {
}

AgoraRenderPerformanceMonitor::~AgoraRenderPerformanceMonitor() {
    delegate_ = nullptr;
}

void AgoraRenderPerformanceMonitor::setDelegate(AgoraRenderPerformanceDelegate* delegate) {
    std::lock_guard<std::mutex> lock(mutex_);
    delegate_ = delegate;
}

void AgoraRenderPerformanceMonitor::setEnabled(bool enabled) {
    std::lock_guard<std::mutex> lock(mutex_);
    enabled_ = enabled;
}

void AgoraRenderPerformanceMonitor::recordFrameReceived() {
    if (!enabled_) return;

    int64_t nowMs = currentTimeMs();

    std::lock_guard<std::mutex> lock(mutex_);
    totalFramesReceived_++;
}

void AgoraRenderPerformanceMonitor::recordFrameRenderedInterval() {
    if (!enabled_) return;

    int64_t nowMs = currentTimeMs();

    std::lock_guard<std::mutex> lock(mutex_);
    
    double frameIntervalMs = 0.0;
    if (last_frame_notify_time_ms_ > 0) {
        frameIntervalMs = static_cast<double>(nowMs - last_frame_notify_time_ms_);
    }
    last_frame_notify_time_ms_ = nowMs;
    last_frame_interval_ms_ = frameIntervalMs;
    
    totalFramesRendered_++;
    reportRawFrameStats();
}

void AgoraRenderPerformanceMonitor::recordRenderDrawCostWithValue(double drawCostMs) {
    if (!enabled_) return;

    std::lock_guard<std::mutex> lock(mutex_);
    last_draw_cost_ms_ = drawCostMs;
}

void AgoraRenderPerformanceMonitor::reportRawFrameStats() {
    std::map<std::string, double> rawStats;
    rawStats["renderFrameIntervalMs"] = last_frame_interval_ms_;
    rawStats["renderDrawCostMs"] = last_draw_cost_ms_;
    rawStats["totalFramesReceived"] = static_cast<double>(totalFramesReceived_);
    rawStats["totalFramesRendered"] = static_cast<double>(totalFramesRendered_);
    
    if (delegate_) {
        delegate_->onRawFrameStats(rawStats);
    }
}

void AgoraRenderPerformanceMonitor::reset() {
    std::lock_guard<std::mutex> lock(mutex_);
    totalFramesReceived_ = 0;
    totalFramesRendered_ = 0;
    last_frame_notify_time_ms_ = 0;
    last_frame_interval_ms_ = 0.0;
    last_draw_cost_ms_ = 0.0;
}

int64_t AgoraRenderPerformanceMonitor::currentTimeMs() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();
}

} // namespace flutter
} // namespace rtc
} // namespace agora

File: android/src/main/java/io/agora/agora_rtc_ng/AgoraRenderPerformanceMonitor.java
package io.agora.agora_rtc_ng;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Delegate interface for performance statistics callbacks.
 */
interface AgoraRenderPerformanceDelegate {
    void onRawFrameStats(Map<String, Object> rawStats);
}

/**
 * Performance monitor for video rendering in Flutter Texture mode.
 */
public class AgoraRenderPerformanceMonitor {
    private static final int MAX_SAMPLE_SIZE = 100;

    private AgoraRenderPerformanceDelegate delegate;
    private boolean enabled = true;

    private long totalFramesReceived = 0;
    private long totalFramesRendered = 0;

    /// Timestamp of last frame notification to Flutter (in milliseconds)
    private long lastFrameNotifyTimeMs = 0;
    /// Timestamp of last frame received (in milliseconds) for render duration calculation
    private long lastFrameReceivedTimeMs = 0;
    /// Last calculated frame interval for immediate reporting
    private double lastFrameIntervalMs = 0;
    /// Last calculated draw cost for immediate reporting
    private double lastDrawCostMs = 0;

    private final Object lock = new Object();

    public AgoraRenderPerformanceMonitor() {
    }
    public void setDelegate(AgoraRenderPerformanceDelegate delegate) {
        synchronized (lock) {
            this.delegate = delegate;
        }
    }

    public void setEnabled(boolean enabled) {
        synchronized (lock) {
            this.enabled = enabled;
        }
    }

    /**
     * Record a frame received event for FPS calculation.
     */
    public void recordFrameReceived() {
        if (!enabled) return;

        long nowMs = System.currentTimeMillis();

        synchronized (lock) {
            lastFrameReceivedTimeMs = nowMs;
            totalFramesReceived++;
        }
    }

    /**
     * Record a frame rendered event (textureFrameAvailable notification).
     * Calculates frame interval from the last notification.
     */
    public void recordFrameRenderedInterval() {
        if (!enabled) return;

        long nowMs = System.currentTimeMillis();

        synchronized (lock) {
            double frameIntervalMs = 0.0;
            if (lastFrameNotifyTimeMs > 0) {
                frameIntervalMs = (double) (nowMs - lastFrameNotifyTimeMs);
            }
            lastFrameNotifyTimeMs = nowMs;
            lastFrameIntervalMs = frameIntervalMs;
            totalFramesRendered++;

            reportRawFrameStats();
        }
    }

    /**
     * Record render draw cost with a pre-calculated value (in milliseconds).
     * This is used when the draw cost is calculated in native code to avoid race conditions.
     */
    public void recordRenderDrawCostWithValue(double drawCostMs) {
        if (!enabled) return;

        synchronized (lock) {
            lastDrawCostMs = drawCostMs;
        }
    }

    private void reportRawFrameStats() {
        Map<String, Object> rawStats = new HashMap<>();
        rawStats.put("renderFrameIntervalMs", lastFrameIntervalMs);
        rawStats.put("renderDrawCostMs", lastDrawCostMs);
        rawStats.put("totalFramesReceived", totalFramesReceived);
        rawStats.put("totalFramesRendered", totalFramesRendered);

        if (delegate != null) {
            delegate.onRawFrameStats(rawStats);
        }
    }

    public void reset() {
        synchronized (lock) {
            totalFramesReceived = 0;
            totalFramesRendered = 0;
            lastFrameNotifyTimeMs = 0;
            lastFrameReceivedTimeMs = 0;
            lastFrameIntervalMs = 0;
            lastDrawCostMs = 0;
        }
    }

    public void dispose() {
        synchronized (lock) {
            delegate = null;
        }
    }
}

